### ISSUE TITLE
Ty + ruff inline hints and line length

### ISSUE THREAD

**Issue by dagl1** [astral-sh/ruff#24267]
### Question

My apologies, I am unsure whether this is a ruff or ty question; additionally I tried looking at tother issues but those appear generally about in-line comments and not inline (editor/ty generated) type and argument annotation.

In my project I utilse some line length before ruff (if it can) reformats the line to adhere to my line length rule as provided in my pyproject.toml. This works great except for strings which is known to be the case. However with ty enabled, I get in-line hints inside my editor (pycharm), which is also great. This however leads to the unfortunate situation that lines **visually** exceed my line length, but this is not picked up upon. 

I am wondering if there is a way around this (I would rather not compromise on either having longer expressions automatically formatted, nor throw away the super handy type hinting). 

<img width="1078" height="127" alt="Image" src="https://github.com/user-attachments/assets/ba6afe2b-1f33-46f2-9768-c2b4c1c44e22" />

Other occurences are picked up correctly (if in strings are highlighted, otherwise automatically fixed). 

Thank you

### Version

ruff 0.15.8 (c2a881584 2026-03-26)

**Comment by MichaReiser** [comment]
I don't think there's much we can do here from a formatter perspective

* Not all inlay hints are valid syntax. That means the formatter can't format the code with inlay hints present. 
* Whether inlay hints are shown depends on your editor settings, and not everyone on your team might have them enabled. That means, taking inlay ints into account when considering the line width may be the right thing for you, but not necessarily for everyone else working with the code (they would see very short lines and not understand why the code doesn't make use of the full line length)
* The formatter is intentionally designed such that formatting decisions don't require information from other files or type information. However, inlay-hints require exactly both. That means changing a function's return type can change the inlay hint at every call site, which in turn could result in that code being reformatted (because the return type is now shorter or longer). Not only would this make formatting much slower, but it also leads to a lot of formatting churn.

We may be able to adjust where and when inlay hints are shown. Although the inlay hints shown in your screenshot look reasonable, I don't see how we could shorten them. 

Could you describe what behavior you would expect?

**Comment by dagl1** [comment]
Hi Micha, 

Those are all great points, I hadn't really considered. 
-- For the first point, I did not imagine anyting to be done **with** the inlay hints, instead that it would (somehow) pick up on what the visual line length would be, and adjust based on that. 
-- Regarding whether others do or do not use type hints (and thus would see shorter lines), is also a great point, although I feel that in and of itself might be acceptable to me (but I doubt that this is worth the technical/developmental effort).
-- Especially this last point is a good one, and I hadn't considered how this would constantly shift lines. I think what I thought I wanted isn't at all what I actually would want.

One thing that I realise I do actually want, is the ability to make any set of arguments or inputs to a function automatically become multiline, without me having to put the trailing comma. I couldn't find any option regarding this, as the documentation seems to allow for the ability to do the opposite with skip-trailing-comma. Would such an option be possible (something like, always-separate-arguments-above-n where in my case i would set it to 2.

``` Python
def some_function(a,b):
```
would automatically act as if there was a trailing comma after the b?

Other than that I realise my initial request is infeasible + not what I actually want in Ruff's behaviour! Thank you for your response with the detailed explanation btw!

Kind regards,

**Comment by MichaReiser** [comment]
> would automatically act as if there was a trailing comma after the b?

Our formatter is intentionally opinionated to reduce discussion about formatting. That's why we keep the number of options limited: every option leads to more discussion.

I think there's some feature that I see in your issue around automatically reflowing user code based on the editor layout (screen size, font size, etc.), which is interesting but likely requires editor integration that goes beyond what editors expose today.

**Comment by dagl1** [comment]
That is fair, I suppose I will have to just accept that I must do some work myself and Ruff won't do everyting for me haha. 
Thanks and looking forward to what other interesting features are going to included!
